### PR TITLE
A major refactoring according to our discussion

### PR DESCRIPTION
**Pull request by wewei**
The following things changed
1. Projection signature `{ decorator, config }`
    * The decorator will only receive the composer, and only return the changed functions
    * The `options` is renamed to `config` as it's another piece of user config.
2. `ProjectionGridCore` take 2 parameters `{ DefaultCell, DefaultHeader }`, so that framework dependent wrappers don't need to pipe a projection to decorate the `composeTds` and `composeThs`.
3. Added the following composer functions
    * `composeHeaderTrs`
    * `composeHeaderThs`
    * `composeDataTrs`
    * `composeDataTds`

Refer to the [Vue config](https://github.com/wewei/vue-projection-grid/blob/renderer/demo/default/app.vue) as an example.

### CODE REVIEW

File: src/builtin-projections/defaults.js [hunk 1]
```diff
@@ -1,37 +1,218 @@
...
+        return this.composeDataTrs(tr);
+      }
+      return [];
+    },
+
+    composeHeaderTrs(tr) {
+      const ths = _.map(
+        columns,
+        column => _.defaults({
+          key: `@${column.name}`,
```
Comment (BigYu) [inline src/builtin-projections/defaults.js]: prefix with “tr-”?

Comment (wewei) [reply src/builtin-projections/defaults.js]: or "@th-“？

File: src/utils/decorator.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    return _.extend({}, events, eventsExt);
+  }
+
+  content(content, contentExt) {
+    const { Component } = contentExt;
+
+    return _.defaults(Component ? {
+      Component,
+      props: _.defaults({ content }, content.props),
+    } : {}, Decorator.create({
```
Comment (haipeng-wang) [inline src/utils/decorator.js]: create(...)

File: src/utils/decorator.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      events: contentExt.events,
```
Comment (wewei) [inline src/utils/decorator.js]:
```js
{
  classes: contentExt.classes,
  style: contentExt.style,
}
```

**Pull request merged**